### PR TITLE
V14: Webhook Management API

### PR DESCRIPTION
**Pull request by bjarnef**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
Added webhooks to Management API and models, which we can generate models for in the new backoffice.
See related PR for new backoffice here: https://github.com/umbraco/Umbraco.CMS.Backoffice/pull/962

### CODE REVIEW

**Comment by bjarnef** [comment]
@Zeegaan @bergmania I have added this while working on https://github.com/umbraco/Umbraco.CMS.Backoffice/pull/962 

I think the CRUD operation methods should take a parameter `Guid userKey` like in `ILanguageService`?

Futhermore should the `IWebhookService` method return `Task<IWebhook>` instead of `Task<Webhook>`?

or maybe `Task<Attempt<IWebhook?, `WebhookOperationStatus>>?

It seems to Update method should also return `Webhook` (or `IWebhook`).

**Comment by georgebid** [comment]
Morning @bjarnef, thanks for PR! Someone on the team will review this soon 😄

**Comment by bjarnef** [comment]
@georgebid it definitely need some more work, but it would be great with some feedback from HQ. Maybe it also requires some additional changes regarding webhooks in `v13/dev` branch.

I know @kjac has worked on the management API, so maybe he could have a look at some point 😎

**Comment by georgebid** [comment]
Ah yes, sorry I totally missed that this was marked as a Draft, but yes don't worry - I am sure someone from HQ will be in touch soon! 🚀

**Comment by bjarnef** [comment]
@elit0451 maybe something you want to review too since you have worked much with the (content) delivery API.
I need to merge latest changes regarding webhooks and later when having the models, I can look at the related backoffice PR 😎

**Comment by bjarnef** [comment]
I think we also need the changes from https://github.com/umbraco/Umbraco-CMS/pull/15180 but it seems these haven't been merged to `v14/dev` branch yet.

**Comment by Zeegaan** [comment]
Should be updated now 😁

**Comment by bjarnef** [comment]
@Zeegaan I have updated that and resolved the conflicts.

I noticed an ambiguous reference because `WebhookResponseModel` exists in `Umbraco.Cms.Core.Models` and also `Umbraco.Cms.Api.Management.ViewModels.Webhook`

**Comment by bjarnef** [comment]
@Zeegaan I think this should be really for review.

I implementered `GetMultipleAsync()` as well, where the `GetByIdsAsync()` in the repository returns a paged model like `GetByAliasAsync()`.

**Comment by bjarnef** [comment]
@kjac maybe you can review this? Then I can continue work in https://github.com/umbraco/Umbraco.CMS.Backoffice/pull/962 😁
I see a few conflicts now though 🙈

**Comment by bjarnef** [comment]
I noticed `CreatedAtAction()` has changed and updated the usage in:
https://github.com/umbraco/Umbraco-CMS/pull/15147/commits/7b48a71f23caf59b7df980347d7f975f56ee1b0f

Should the `routeValues` be nullable? and perhaps an overload as before accepting `Guid` as identifier?

**Comment by Zeegaan** [comment]
@bjarnef We are aware of this, and we have set aside sprint time to review, but we are really busy with all kinds of v14 stuff 🏃‍♂️ 

I will get to this as soon as I am able 😁

**Comment by bjarnef** [comment]
@Zeegaan yeah, I know.. and seems there are lots to do.
But I guess it need this sooner or later 😁

and would love to help with the frontend (in backoffice) webhooks workspace, but I neeed to generate the models from `v14/dev` branch to continue that work 😅

File: src/Umbraco.Cms.Api.Management/Controllers/Webhook/ByKeyWebhookController.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    [HttpGet("{id:guid}")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(WebhookResponseModel), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
+    public async Task<IActionResult> ByKey(Guid id)
+    {
+        IWebhook? webhook = await _webhookService.GetAsync(id);
+        if (webhook == null)
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/Controllers/Webhook/ByKeyWebhookController.cs:31]:
```suggestion
        if (webhook is null)
```

File: src/Umbraco.Cms.Api.Management/Controllers/Webhook/ByKeyWebhookController.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        {
+            return WebhookOperationStatusResult(WebhookOperationStatus.NotFound);
+        }
+
+        WebhookResponseModel model = _umbracoMapper.Map<WebhookResponseModel>(webhook)!;
+        return await Task.FromResult(Ok(model));
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/Controllers/Webhook/ByKeyWebhookController.cs:37]: Should be good to just return the model as we're using an `await` in this method 😁 
```suggestion
        return Ok(model);
```

File: src/Umbraco.Cms.Api.Management/Controllers/Webhook/DeleteWebhookController.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    [HttpDelete($"{{{nameof(id)}}}")]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
+    [ProducesResponseType(StatusCodes.Status200OK)]
+    public async Task<IActionResult> Delete(Guid id)
+    {
+        Attempt<IWebhook?, WebhookOperationStatus> result = await _webhookService.DeleteAsync(id); //, CurrentUserKey(_backOfficeSecurityAccessor));
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/Controllers/Webhook/DeleteWebhookController.cs:34]:
```suggestion
        Attempt<IWebhook?, WebhookOperationStatus> result = await _webhookService.DeleteAsync(id);
```

File: src/Umbraco.Cms.Api.Management/Controllers/Webhook/CreateWebhookController.cs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    [HttpPost]
+    [MapToApiVersion("1.0")]
+    [ProducesResponseType(StatusCodes.Status201Created)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(typeof(ProblemDetails), StatusCodes.Status404NotFound)]
+    public async Task<IActionResult> Create(CreateWebhookRequestModel createWebhookRequestModel)
+    {
+        Core.Models.Webhook created = _umbracoMapper.Map<Core.Models.Webhook>(createWebhookRequestModel)!;
+
+        Attempt<IWebhook, WebhookOperationStatus> result = await _webhookService.CreateAsync(created); //, CurrentUserKey(_backOfficeSecurityAccessor));
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/Controllers/Webhook/CreateWebhookController.cs:44]: I'd rather have TODOs in the service telling us to implement audit logging for this 😁 
```suggestion
        Attempt<IWebhook, WebhookOperationStatus> result = await _webhookService.CreateAsync(created);
```

File: src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs [hunk 5]
```diff
@@ -0,0 +1,6 @@
+namespace Umbraco.Cms.Api.Management.ViewModels.Webhook;
+
+public class CreateWebhookRequestModel : WebhookModelBase
+{
+    
+}
```
Comment (Zeegaan) [inline src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:6]: This model should have a nullable key, and we should be creating the webhook with said Id (Key in backend terms 🤪 )if its not null.

(This is for use cases like offline-mode support in the future) 👍 
```suggestion
{
    public Guid? Id { get; set; }
}
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Sorry for this taking a long time, finally got around to it 😁 
Have a few comments, but overall looks splendid! 💪

**Comment by bjarnef** [comment]
Regarding the `CurrentUserKey()` and audits of the the webhook e.g. was created from backoffice or from code is need further adjustments. IIRC languages have this, but not audit in backoffice UI.

Maybe the webhook model need some adjustments as well. I haven't checked with the latest changes with was added/adjusted in v13.

The webhook inherits `TEntity` IIRC, but the UI doesn't allow setting a werbhook friendly name yet I think. In case this will be added it could use the header name field in backoffice, but not if it is optional as that requires a value.
Maybe only in v14 later and then webhook model with a name property as well.

**Comment by Zeegaan** [comment]
Yea I think a friendly name at some point would be nice, but I think we should aim for feature parity for now, and then we can always sprinkle stuff on top later 😁 

I will give this a test for now, and come back if the models needs adjustment 👍

File: src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+namespace Umbraco.Cms.Api.Management.ViewModels.Webhook;
+
+public class CreateWebhookRequestModel : WebhookModelBase
+{
+    public Guid? Id { get; set; }
```
Comment (bjarnef) [inline src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: @Zeegaan does `UpdateWebhookRequestModel` also need this and `WebhookResponseModel`?

Should it be part of `WebhookModelBase` instead then?

Comment (Zeegaan) [reply src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: No just the CreateModel! 😁
It's basically just so you can specify an Id upon creation 👍

Comment (bjarnef) [reply src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: Okay, but we should probably add it to `WebhookResponseModel` model as well then - just like with datatype:
https://github.com/umbraco/Umbraco-CMS/blob/268babf8895e0088591ebb4e4efa15112749baa3/src/Umbraco.Cms.Api.Management/ViewModels/DataType/DataTypeResponseModel.cs#L5

Comment (bjarnef) [reply src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: I added `Guid` to `WebhookResponseModel`:
https://github.com/umbraco/Umbraco-CMS/pull/15147/commits/500513870b6986e147fb89a8e17a944bf33f2d7c

I noticed other response models inherits from `NamedItemResponseModelBase` which inherits from `ItemResponseModelBase` and has `Guid Id`.

E.g. `MemberGroupItemResponseModel` inherits from `NamedItemResponseModelBase`.

Comment (bjarnef) [reply src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: I guess `WebhookResponseModel` can inherits from `NamedItemResponseModelBase` as webhook inherits `TEntity` or from `ItemResponseModelBase` ?

Comment (Zeegaan) [reply src/Umbraco.Cms.Api.Management/ViewModels/Webhook/CreateWebhookRequestModel.cs:5]: `ItemResponseModelBase` are for the item endpoints, so no more inheritance please 🙈

**Comment by Zeegaan** [comment]
@bjarnef I've added some missing mapping and the new auth policies, but works like a charm now 😁 
Will merge when the tests pass 🚀

**Comment by bjarnef** [comment]
@Zeegaan added in https://github.com/umbraco/Umbraco-CMS/pull/15802

**Pull request merged**